### ISSUE TITLE
Multiplayer support

### ISSUE THREAD

**Issue by techniponi** [RSDuck/vba-next-switch#25]
(I'm moving this from the comments on #22)

Several GBA games support multiplayer, for example:
* Kirby & the Amazing Mirror
* The Legend of Zelda: Four Swords
* Super Mario Advance series (Mario Bros.)

These games especially have a very "splitscreen-like" experience, and having them running in a splitscreen configuration (by emulating 2-4 simultaneous GBAs) would be a perfectly viable gameplay experience for these titles. Each GBA would simply be associated to player 1-4's input on the Switch (this would benefit from #23, or at least single sideways joycon controls).

The one question for implementation - how would you manage save files?
I would personally suggest one of these two implementations:
* A unique save file for each controller slot
  * This might present issues if players ever switch slots (save data would need to be manually renamed, unless there was a "move save slot" option somewhere)
* A rudimentary 'profile' system in which each controller selects their desired savefile before emulation starts.
  * Obviously slightly harder to implement, but much more practical for couch multiplayer setups.
  * I might recommend using something like color-coding or character icons as a substitute for actual player names (so everyone doesn't have to take 5 minutes typing their names)

---

Netplay, online, or NFC/wireless multiplayer might be far more challenging to implement. The GBA's link cable expects zero latency, and most games will crash or freeze when the link cable desyncs. The only viable alternative I can imagine is to extend the splitscreen functionality in the style of RetroArch/Dolphin/console netplay: button input sync. Rather than broadcasting the link cable data over the internet, each console would simultaneously emulate both GBAs and would synchronize the save data with each other, and would simply broadcast and sync button inputs.

This still has issues with desync and latency of course, so I still don't know if netplay is worth the trouble.

**Comment by fennectech** [comment]
I support this.  Net play would be MUCH harder and really isn’t worth the time. I don’t think anyone has gotten it to work.    However. Emulating bolth GBAs in the cloud and streaming it live with a video link would actually work. And simply syncing the battery files at the end of the play session to work.     However an implementation like this would mandate a paid service that would be rather unpopular (not like unliked but unused) becuase many people don’t play GBA.  Local multiplayer is the best way to go.   Like it was with the GBA.    If people want to multiplayer there are other options and the switch is not well suited to this.

**Comment by fennectech** [comment]
But with local multiplayer one switch could emulate BOLTH game boys and stream the game to the other switch (should be plenty powerful enough to do this) and sync the battery files over the local link at the end of the play session.  This would be tough to implement but it could work quite well.     Glitches in the stream would only cause visual issues and audio issues.

**Comment by fennectech** [comment]
And in a full disconnect bolth switches could simply try to reconnect. And if it fails.  The host switch could could dump the other players save file to the sdcard and let the players sort it out.      There would be a page on this in the manual    And make sure multiple save files could be saved. (Ie with a date-time-link.sav) with instructions on how to get the file to the other player with a PC

**Comment by fennectech** [comment]
Tell me if my solution is too ambitious  I got the idea from steam link.    With proper optimization the switch could do all of this.

**Comment by techniponi** [comment]
It's very ambitious, essentially you'd have to figure out how to encode the video in a way that doesn't interfere with the performance of the game, and send it to the other switch to be displayed with as little latency as possible.

I absolutely believe it would be possible, but it would require research into how to transfer video data via local multiplayer. Honestly, doing it over LAN might actually be easier.

I still personally think splitscreen would be more than enough.

**Comment by fennectech** [comment]
With such a low resolution and such a high bandwidth available      There are already plenty of formats that have minimal compression.   We could easily drop the frame rate a bit.    Doesn’t the tegra have a hardware codec?

**Comment by fennectech** [comment]
Yeah.  Tegra X1 has a HEVC codec suitable for the process.   https://developer.nvidia.com/content/tegra-x1

**Comment by fennectech** [comment]
Switch is outside of my realm.  I could not dream to implement this but with the tegra X1’s hardware codec we don’t have to implement the codec in software meaning it cuts out a helluva lot of overhead.   https://international.download.nvidia.com/pdf/tegra/Tegra-X1-whitepaper-v1.0.pdf

**Comment by fennectech** [comment]
How does my methodology sound.   Any holes in it?     Anything need clarification?   Am I looking at this all wrong?

**Comment by RSDuck** [comment]
We don't even have GPU access yet, so there's a clear no to this one. Besides I think the whole concept of streaming would be a bit of an overkill.

The problem with splitscreen is that the emulator architecture is built in a way that allows only one emulator at the time. So probably would need some work on the core emulator to get this one done. But it's far more realistic.

Otherwise LAN multiplayer would probably be the easiest to implement, because it can be integrated from other variants of VBA, although I don't know good it's going to end up.

**Comment by fennectech** [comment]
So we have to wait for Libnx mature

**Comment by RSDuck** [comment]
there is support for hosting a local wireless in the Switch, which is used quite often in games with local multiplayer(Mario Kart, …). 

Unfortunately libnx currently doesn't support interfacing the necessary services, so you would have to be in the same LAN.

**Comment by techniponi** [comment]
If it's possible to do it within the same LAN, **in theory** some or all of the same code could be reused and adapted to the local multiplayer network once libnx has support for it. Plus, having the feature *at all* (even in limited form) is always nicer than nothing.

**Comment by fennectech** [comment]
Yeah.    It’s definitely something that could be viable in the future though if/when libnx is extended to support these features.
